pkg/chrootarchive/archive_freebsd.go

```diff
@@ -0,0 +1,237 @@
+		}
+	}()
+
+	if err = encodeJson(options, optionsW); err != nil {
+		return nil, errors.Wrap(err, "json encode to pipe failed")
+	}
+
+	return r, nil
+}
+
+func doUnpackLayer(root string, layer io.Reader, options *archive.TarOptions) (chan (unpackLayerResult), error) {
+	done := make(chan unpackLayerResult, 1)
+	optionsR, optionsW, err := os.Pipe()
+	if err != nil {
+		return done, errors.Wrap(err, "could not create a pipe")
+	}
+	defer optionsW.Close()
+	defer optionsR.Close()
+	buffer := bytes.NewBuffer(nil)
+
+	cmd := reexec.Command(unpackLayerCmd, root)
+	cmd.Stdin = layer
+	cmd.Stderr = buffer
+	cmd.ExtraFiles = []*os.File{
+		optionsR,
+	}
+
+	if err = cmd.Start(); err != nil {
+		return done, errors.Wrap(err, "re-exec error")
+	}
+
+	if err = encodeJson(options, optionsW); err != nil {
+		return done, errors.Wrap(err, "json encode to pipe failed")
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return done, errors.Wrap(err, buffer.String())
+	}
+
+	result, err := decodeUnpackLayerResult(buffer)
+
+	if err != nil {
+		return done, errors.Wrap(err, "json decoding error")
+	}
+
+	done <- result
+
+	return done, nil
+}
+
+func unpackInChroot() {
+	relDest := os.Args[1]
+	root := os.Args[2]
+	options := recvOptions()
+
+	if err := syscall.Chroot(root); err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrap(err, "could not chroot"))
+		os.Exit(1)
+	}
+	if err := archive.Unpack(os.Stdin, relDest, &options); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+}
+
+func packInChroot() {
+	relSrc := os.Args[1]
+	root := os.Args[2]
+	options := recvOptions()
+
+	tb, err := archive.NewTarballer(relSrc, &options)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrap(err, "error processing tar file"))
+		os.Exit(1)
+	}
+	err = syscall.Chroot(root)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrap(err, "could not chroot"))
+		os.Exit(1)
+	}
+
+	go tb.Do()
+
+	io.Copy(os.Stdout, tb.Reader())
+}
+
+func unpackLayerInChroot() {
+	root := os.Args[1]
+	options := recvOptions()
+
+	if err := syscall.Chroot(root); err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrap(err, "could not chroot"))
+		os.Exit(1)
+	}
+
+	// We need to be able to set any perms
+	_ = unix.Umask(0)
+	size, err := archive.UnpackLayer("/", os.Stdin, &options)
+
+	encodeJson(unpackLayerResult{layerSize: size, err: err}, os.Stdout)
+}
+
+func pipeOptions(options *archive.TarOptions) (*os.File, *os.File, error) {
+	optionsR, optionsW, err := os.Pipe()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not create a pipe")
+	}
+	return optionsR, optionsW, encodeJson(options, optionsW)
+}
+
+func recvOptions() archive.TarOptions {
+	optionsPipe := os.NewFile(3, "tar-options")
+	if optionsPipe == nil {
+		fmt.Fprintln(os.Stderr, "could not read tar options from the pipe")
+		os.Exit(1)
+	}
+	defer optionsPipe.Close()
+	options, err := decodeOptions(optionsPipe)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrap(err, "could not decode tar options"))
```

Comment:
Very quick blurb (about to head into a call); I think all of these errors (also elsewhere in this file) are part of the  reexec's and won't be unwrapped by anyone, other than being printed on the command-line.

I would suggest using stdlib  `error` / `fmt.Errorf()` for these if we don't need the added features of pkg/errors.

For this specific case, I guess it doesn't even need to be an error, as we're just printing, so this could be a `fmt.Fprintf` (with a newline in the format string)

Reply:
Thanks, fixed.